File: ngmath/src/lib/gridpack/csagrid/csa3xs.f
C                                                                      
C                Copyright (C)  2000
C        University Corporation for Atmospheric Research
C                All Rights Reserved
C
C The use of this Software is governed by a License Agreement.
C
      SUBROUTINE CSA3XS (NI,XI,UI,WTS,KNOTS,SSMTH,NDERIV,NXO,NYO,NZO,
     +                  XO,YO,ZO,UO,NWRK,WORK,IER)
C
      DIMENSION XI(3,NI),UI(NI),WTS(*),KNOTS(3),XO(NXO),YO(NYO),
     +          ZO(NZO),UO(NXO,NYO,NZO),WORK(NWRK),NDERIV(3)
      DIMENSION XMN(3),XMX(3),XV(3)
C
C  Check on the number of knots.
C
      NTOT = KNOTS(1)*KNOTS(2)*KNOTS(3)
C
      DO 20 I=1,3
        IF (KNOTS(I) .LT. 4) THEN
          CALL CFAERR (202,' CSA3XS - must have at least four knots in e
     +very coordinate direction',68)       
          IER = 202
          RETURN
        ENDIF
   20 CONTINUE
C
C  Check on the size of the workspace.
C
      IF (NWRK .LT. NTOT*(NTOT+3)) THEN
        CALL CFAERR (203,' CSA3XS - workspace too small',29)
        IER = 202
        RETURN
      ENDIF
C
C  Calculate the min and max for the knots as the minimum value of
C  the input data and output data and the maximum of the input and
C  output data.
C
      DO 30 J=1,3
        XMN(J) = XI(J,1)
        XMX(J) = XI(J,1)
        DO 10 I=2,NI
          XMN(J) = MIN(XMN(J),XI(J,I))
          XMX(J) = MAX(XMX(J),XI(J,I))
   10   CONTINUE
   30 CONTINUE
      XMN(1) = MIN(XMN(1),XO(1))
      XMX(1) = MAX(XMX(1),XO(NXO))
      XMN(2) = MIN(XMN(2),YO(1))
      XMX(2) = MAX(XMX(2),YO(NYO))
      XMN(3) = MIN(XMN(3),ZO(1))
      XMX(3) = MAX(XMX(3),ZO(NZO))
C
C  Find the coefficients.
C
      CALL SPLCW(3,XI,3,UI,WTS,NI,XMN,XMX,KNOTS,SSMTH,WORK,NTOT,       
     +           WORK(NTOT+1),NWRK-NTOT,IERR)
      IF (IERR .NE. 0) RETURN
C
C  Calculate the approximated values (coefficients are stored at
C  the beginnig of WORK).
C
      DO 60 I=1,NXO
        XV(1) = XO(I)
        DO 40 J=1,NYO
          XV(2) = YO(J)
          DO 50 K=1,NZO
            XV(3) = ZO(K)
            UO(I,J,K) = SPLDE(3,XV,NDERIV,WORK,XMN,XMX,KNOTS,IER)
            IF (IERR .NE. 0) RETURN
   50     CONTINUE
   40   CONTINUE
   60 CONTINUE
C
      RETURN
      END

File: ngmath/man/csagrid/csa3xs.m
.\"
.\"	$Id: csa3xs.m,v 1.4 2008-07-27 03:35:34 haley Exp $
.\"
.TH CSA3XS 3NCARG "January 1999" UNIX "NCAR GRAPHICS"
.SH NAME
CSA3XS - cubic spline approximation, expanded entry for three-dimensional input
.SH SYNOPSIS
CALL CSA3XS (NI, XI, UI, WTS, KNOTS, SMTH, NDERIV, 
.br
             NXO, ,NYO, NZO, XO, YO, ZO, UO, NWRK, 
.br
             WORK, IER)
.SH DESCRIPTION
.IP NI 12
(integer,input) The number of input data points. It must be that NI .gt. 3 
and, depending on the size of KNOTS below, NI may have to be larger.
.IP XI 12
(real, input) An array containing the X - Y - Z coordinates of the 
input data points.  XI is dimensioned for 3 x NI.  XI(1,L) is the X 
coordinate, XI(2,L) is the Y coordinate, and XI(2,L) is the Z coordinate 
for the input domain for L=1,NI.
.IP UI 12
(real, input) An array dimensioned for NI
containing function values at the input XI values, 
that is, UI(L) is the value of the input function at XI(L) for L=1,NI.
through the input function values.
.IP WTS 12
(real, input) An array dimensioned for NI containing weights for the UI 
values at the input XI values, that is, WTS(L) is a weight for the 
value of UI(L) for L=1,NI.  If you do not desire to weight the input 
UI values, then set WTS(1) to -1.  The weights in the WTS array are 
relative and may be set to any non-negative value.  When CSA3XS is called,
the weights are summed and the individual weights are normalized
so that the weight sum is unity.
.IP KNOTS 12
(integer, input) The number 
of knots to be used in constructing the approximation
spline.  KNOTS is dimensioned for 3 and provides the number of knots to be
used in the X, Y, and  Z directions.  Both KNOTS(I) must be at
least 4 for I=1,3.  The larger the values for KNOTS, the closer 
the approximated curve will come to passing through the input 
function values.
.IP SMTH 12
(real, input) A parameter that controls extrapolation into
data sparse regions.  If SMTH is zero, then nothing special 
is done in data sparse regions.  A good first choice for SMTH is 1.
.IP NDERIV 12
(integer, input) An array dimensioned for 3 that specifies, for each 
coordinate,  whether you want functional values (=0),
first derivative values (=1), or second derivative values (=2).
.IP NXO 12
(integer, input) The number of X coordinate values in the output grid.
.IP NYO 12
(integer, input) The number of Y coordinate values in the output grid.
.IP NZO 12
(integer, input) The number of Z coordinate values in the output grid.
.IP XO 12
(real, input) An array dimensioned for NXO 
containing the X coordinates of the output spline.
.IP YO 12
(real, input) An array dimensioned for NYO
containing the Y coordinates of the output spline.
.IP ZO 12
(real, input) An array dimensioned for NZO
containing the Z coordinates of the output spline.
.IP UO 12
(real, output) An array dimensioned for NXO x NYO x NZO 
containing the calculated function values for the
output function.  UO(I,J,K) is the calculated functional value
at (XO(I), YO(J), ZO(K)) for I=1,NXO and J=1,NYO and K=1,NZO.
.IP NWRK 12 
(integer, input) The size of the WORK array.  NWRK must be at least
NK * (NK+3) where NK = KNOTS(1) * KNOTS(2) * KNOTS(3).
.IP WORK 12
(real, input) A work array dimensioned for NWRK.
.IP IER 12
(integer, output) An error return value.  If IER is returned as 0, then
no errors were detected. If IER is non-zero, then refer to the man
page for csagrid_errors for details.
.SH USAGE
CSA3XS is called to find an approximating cubic spline for three-dimensional
input data.  CSA3XS is called if you want to weight the input
data values, calculate derivatives, or handle data sparse areas specially.
If you do not want to do any of these three things, then use CSA3S.
.SH ACCESS
To use CSA3XS, load the NCAR Graphics library ngmath.
.SH SEE ALSO
csagrid,
csa3s,
csa3ls,
csa3lxs
.sp
Complete documentation for Csagrid is available at URL
.br
http://ngwww.ucar.edu/ngdoc/ng/ngmath/csagrid/csahome.html
.SH COPYRIGHT
Copyright (C) 2000
.br
University Corporation for Atmospheric Research
.br

The use of this Software is governed by a License Agreement.
